File: scripts/test-vision-analysis.ts
/**
 * Vision/Image Analysis Test Script for Midas API
 *
 * Tests vision capabilities of Claude Sonnet-4, GPT-4o, and Gemini models
 *
 * Usage:
 *   # Test with a local image file
 *   npx ts-node scripts/test-vision-analysis.ts --image path/to/image.jpg
 *
 *   # Test with a URL image
 *   npx ts-node scripts/test-vision-analysis.ts --url https://example.com/image.jpg
 *
 *   # Test with a specific model
 *   npx ts-node scripts/test-vision-analysis.ts --image photo.jpg --model "GPT 4o"
 *
 *   # Verbose output
 *   npx ts-node scripts/test-vision-analysis.ts --image photo.jpg --verbose
 *
 *   # Generate a test pattern image
 *   npx ts-node scripts/test-vision-analysis.ts --generate-test-image
 */

import * as fs from 'fs';
import * as path from 'path';

// Types
interface VisionTestResult {
  model: string;
  success: boolean;
  responseTime: number;
  response?: string;
  error?: string;
  statusCode?: number;
  tokensUsed?: number;
}

interface ModelConfig {
  name: string;
  deploymentName: string;
  supportsVision: boolean;
}

// Configuration
const ENDPOINT = 'https://midas.ai.bosch.com/ss1/api/v2/llm/completions';
const API_KEY = process.env.REACT_APP_AZURE_API_KEY || process.env.MIDAS_API_KEY || '';

// Models to test (vision-capable)
const MODELS: ModelConfig[] = [
  { name: 'Claude Sonnet-4', deploymentName: 'Claude-Sonnet-4', supportsVision: true },
  { name: 'GPT 4o', deploymentName: 'GPT 4o', supportsVision: true },
  { name: 'GPT 4.1', deploymentName: 'GPT 4.1', supportsVision: true },
  { name: 'Gemini 2.5 Pro', deploymentName: 'Gemini-2.5-pro', supportsVision: true },
  { name: 'Gemini 2.0 Flash', deploymentName: 'Gemini-2.0-flash', supportsVision: true },
];

// Parse command line arguments
const args = process.argv.slice(2);
const verbose = args.includes('--verbose') || args.includes('-v');
const imagePath = args.find(arg => arg.startsWith('--image='))?.split('=')[1] ||
                  args[args.indexOf('--image') + 1];
const imageUrl = args.find(arg => arg.startsWith('--url='))?.split('=')[1] ||
                 args[args.indexOf('--url') + 1];
const specificModel = args.find(arg => arg.startsWith('--model='))?.split('=')[1];
const generateTest = args.includes('--generate-test-image');
const apiKeyArg = args.find(arg => arg.startsWith('--api-key='))?.split('=')[1];
const finalApiKey = apiKeyArg || API_KEY;

// Colors for console output
const colors = {
  reset: '\x1b[0m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
  gray: '\x1b[90m',
};

function log(message: string, color: keyof typeof colors = 'reset') {
  console.log(`${colors[color]}${message}${colors.reset}`);
}

function logSection(title: string) {
  console.log('\n' + '='.repeat(80));
  log(title, 'cyan');
  console.log('='.repeat(80) + '\n');
}

/**
 * Convert image file to base64
 */
function imageToBase64(filePath: string): string {
  try {
    const imageBuffer = fs.readFileSync(filePath);
    return imageBuffer.toString('base64');
  } catch (error) {
    throw new Error(`Failed to read image file: ${error}`);
  }
}

/**
 * Download image from URL and convert to base64
 */
async function downloadImageToBase64(url: string): Promise<string> {
  try {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`Failed to download image: ${response.status} ${response.statusText}`);
    }
    const buffer = await response.arrayBuffer();
    return Buffer.from(buffer).toString('base64');
  } catch (error) {
    throw new Error(`Failed to download image from URL: ${error}`);
  }
}

/**
 * Generate a simple test image (colored squares pattern)
 * Returns base64 encoded PNG
 */
function generateTestImage(): string {
  // Create a simple test pattern as SVG, then convert to base64
  // This is a simple colored square pattern
  const svg = `<?xml version="1.0" encoding="UTF-8"?>
<svg width="400" height="300" xmlns="http://www.w3.org/2000/svg">
  <!-- Background -->
  <rect width="400" height="300" fill="#f0f0f0"/>

  <!-- Colored squares -->
  <rect x="50" y="50" width="80" height="80" fill="#FF6B6B" stroke="#000" stroke-width="2"/>
  <rect x="160" y="50" width="80" height="80" fill="#4ECDC4" stroke="#000" stroke-width="2"/>
  <rect x="270" y="50" width="80" height="80" fill="#45B7D1" stroke="#000" stroke-width="2"/>

  <rect x="105" y="160" width="80" height="80" fill="#FFA07A" stroke="#000" stroke-width="2"/>
  <rect x="215" y="160" width="80" height="80" fill="#98D8C8" stroke="#000" stroke-width="2"/>

  <!-- Text -->
  <text x="200" y="30" font-family="Arial" font-size="24" fill="#333" text-anchor="middle">
    Test Pattern
  </text>
</svg>`;

  // Convert SVG to base64
  return Buffer.from(svg).toString('base64');
}

/**
 * Save generated test image to file
 */
function saveTestImage(base64Data: string, outputPath: string) {
  const buffer = Buffer.from(base64Data, 'base64');
  fs.writeFileSync(outputPath, buffer);
  log(`✅ Test image saved to: ${outputPath}`, 'green');
}

/**
 * Test vision analysis with a single model
 */
async function testVisionModel(
  model: ModelConfig,
  imageBase64: string,
  imageType: string = 'jpeg'
): Promise<VisionTestResult> {
  const startTime = Date.now();

  try {
    log(`  Testing ${model.name}...`, 'gray');

    // Prepare the vision request (OpenAI format)
    const request = {
      model: model.deploymentName,
      messages: [
        {
          role: 'user',
          content: [
            {
              type: 'text',
              text: `Analyze this image in detail. Describe:
1. What you see in the image (objects, people, scene)
2. Colors and composition
3. Lighting conditions
4. Overall mood or atmosphere
5. Any text visible in the image

Be specific and detailed.`
            },
            {
              type: 'image_url',
              image_url: {
                url: `data:image/${imageType};base64,${imageBase64}`,
                detail: 'high' // Request high detail analysis
              }
            }
          ]
        }
      ],
      max_tokens: 500,
      temperature: 0.3,
    };

    const response = await fetch(ENDPOINT, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...(finalApiKey ? { 'Authorization': `Bearer ${finalApiKey}` } : {}),
      },
      body: JSON.stringify(request),
    });

    const responseTime = Date.now() - startTime;
    const data = await response.json();

    if (!response.ok) {
      return {
        model: model.name,
        success: false,
        responseTime,
        statusCode: response.status,
        error: data.error?.message || data.message || JSON.stringify(data),
      };
    }

    // Extract response text (handle both wrapped and unwrapped responses)
    const wrappedData = data.data || data;
    const content = wrappedData.choices?.[0]?.message?.content || '';
    const tokensUsed = wrappedData.usage?.total_tokens;

    return {
      model: model.name,
      success: true,
      responseTime,
      statusCode: response.status,
      response: content,
      tokensUsed,
    };
  } catch (error) {
    const responseTime = Date.now() - startTime;
    return {
      model: model.name,
      success: false,
      responseTime,
      error: error instanceof Error ? error.message : String(error),
    };
  }
}

/**
 * Main test function
 */
async function runVisionTests() {
  logSection('🔍 Midas API Vision/Image Analysis Test');

  // Handle test image generation
  if (generateTest) {
    logSection('Generating Test Image');
    const testImageBase64 = generateTestImage();
    const outputPath = path.join(__dirname, 'test-pattern.svg');
    saveTestImage(testImageBase64, outputPath);
    log('\nYou can now test with this image:', 'yellow');
    log(`  npx ts-node scripts/test-vision-analysis.ts --image ${outputPath}`, 'cyan');
    return;
  }

  // Validate inputs
  if (!imagePath && !imageUrl) {
    log('❌ Error: No image source provided!', 'red');
    log('\nUsage:', 'yellow');
    log('  --image path/to/image.jpg    Load image from file', 'cyan');
    log('  --url https://example.com/image.jpg    Load image from URL', 'cyan');
    log('  --generate-test-image         Generate a test pattern SVG', 'cyan');
    log('  --model "GPT 4o"             Test specific model only', 'cyan');
    log('  --verbose                     Show full responses', 'cyan');
    process.exit(1);
  }

  // Load image
  let imageBase64: string;
  let imageType: string = 'jpeg';

  try {
    if (imagePath) {
      log(`Loading image from file: ${imagePath}`, 'gray');
      imageBase64 = imageToBase64(imagePath);

      // Detect image type from extension
      const ext = path.extname(imagePath).toLowerCase();
      if (ext === '.png') imageType = 'png';
      else if (ext === '.gif') imageType = 'gif';
      else if (ext === '.webp') imageType = 'webp';
      else if (ext === '.svg') imageType = 'svg+xml';

      const fileSizeKB = Math.round(Buffer.from(imageBase64, 'base64').length / 1024);
      log(`✅ Image loaded: ${fileSizeKB} KB (${imageType})`, 'green');
    } else if (imageUrl) {
      log(`Downloading image from URL: ${imageUrl}`, 'gray');
      imageBase64 = await downloadImageToBase64(imageUrl);
      const fileSizeKB = Math.round(Buffer.from(imageBase64, 'base64').length / 1024);
      log(`✅ Image downloaded: ${fileSizeKB} KB`, 'green');
    } else {
      throw new Error('No image source provided');
    }
  } catch (error) {
    log(`❌ Failed to load image: ${error}`, 'red');
    process.exit(1);
  }

  // Filter models if specific model requested
  const modelsToTest = specificModel
    ? MODELS.filter(m => m.name === specificModel || m.deploymentName === specificModel)
    : MODELS;

  if (modelsToTest.length === 0) {
    log(`❌ Model not found: ${specificModel}`, 'red');
    log('Available models:', 'yellow');
    MODELS.forEach(m => log(`  - ${m.name}`, 'cyan'));
    process.exit(1);
  }

  // Run tests
  logSection('Running Vision Analysis Tests');
  const results: VisionTestResult[] = [];
  let successCount = 0;
  let failCount = 0;

  for (const model of modelsToTest) {
    const result = await testVisionModel(model, imageBase64, imageType);
    results.push(result);

    if (result.success) {
      successCount++;
      log(`  ✅ ${result.model} - ${result.responseTime}ms`, 'green');
      if (result.tokensUsed) {
        log(`     Tokens used: ${result.tokensUsed}`, 'gray');
      }
      if (verbose && result.response) {
        log(`\n     Response:`, 'cyan');
        log(`     ${result.response.substring(0, 500)}${result.response.length > 500 ? '...' : ''}`, 'gray');
        log('', 'reset');
      }
    } else {
      failCount++;
      log(`  ❌ ${result.model} - ${result.responseTime}ms`, 'red');
      if (result.statusCode) {
        log(`     Status: ${result.statusCode}`, 'red');
      }
      if (result.error) {
        log(`     Error: ${result.error.substring(0, 200)}`, 'red');
      }
    }

    // Small delay to avoid rate limiting
    await new Promise(resolve => setTimeout(resolve, 1000));
  }

  // Summary
  logSection('Test Summary');
  log(`Total tests: ${results.length}`, 'cyan');
  log(`✅ Successful: ${successCount}`, 'green');
  log(`❌ Failed: ${failCount}`, 'red');
  if (results.length > 0) {
    log(`Success rate: ${((successCount / results.length) * 100).toFixed(1)}%`, 'cyan');
  }

  // Average response time for successful tests
  const successfulResults = results.filter(r => r.success);
  if (successfulResults.length > 0) {
    const avgTime = successfulResults.reduce((sum, r) => sum + r.responseTime, 0) / successfulResults.length;
    log(`Average response time: ${Math.round(avgTime)}ms`, 'cyan');
  }

  // Save results to file
  const resultsPath = path.join(__dirname, 'vision-test-results.json');
  fs.writeFileSync(
    resultsPath,
    JSON.stringify(
      {
        timestamp: new Date().toISOString(),
        imageSource: imagePath || imageUrl,
        imageType,
        totalTests: results.length,
        successCount,
        failCount,
        results: results.map(r => ({
          ...r,
          // Truncate response in saved file for readability
          response: r.response ? r.response.substring(0, 500) : undefined,
        })),
      },
      null,
      2
    )
  );
  log(`\n📝 Results saved to: ${resultsPath}`, 'blue');

  // Show full responses if not in verbose mode
  if (!verbose && successCount > 0) {
    logSection('Sample Responses (first 200 chars each)');
    results
      .filter(r => r.success && r.response)
      .forEach(r => {
        log(`\n${r.model}:`, 'cyan');
        log(r.response!.substring(0, 200) + '...', 'gray');
      });
    log('\n💡 Use --verbose flag to see full responses', 'yellow');
  }

  console.log('\n');
}

// Run tests
runVisionTests().catch(error => {
  log('Fatal error:', 'red');
  console.error(error);
  process.exit(1);
});

File: scripts/test-vision-analysis.py
#!/usr/bin/env python3
"""
Vision/Image Analysis Test Script for Midas API (Python version)

Tests vision capabilities of Claude Sonnet-4, GPT-4o, and Gemini models

Usage:
    python scripts/test-vision-analysis.py --image path/to/image.jpg
    python scripts/test-vision-analysis.py --url https://example.com/image.jpg
    python scripts/test-vision-analysis.py --image photo.jpg --model "GPT 4o"
    python scripts/test-vision-analysis.py --generate-test-image
    python scripts/test-vision-analysis.py --image photo.jpg --verbose
"""

import os
import sys
import json
import base64
import time
import argparse
from datetime import datetime
from pathlib import Path
from typing import Optional, Dict, Any, List
import requests

# Configuration
ENDPOINT = "https://midas.ai.bosch.com/ss1/api/v2/llm/completions"
API_KEY = os.environ.get("REACT_APP_AZURE_API_KEY") or os.environ.get("MIDAS_API_KEY") or ""

# Models to test (vision-capable)
VISION_MODELS = [
    {"name": "Claude Sonnet-4", "deployment": "Claude-Sonnet-4"},
    {"name": "GPT 4o", "deployment": "GPT 4o"},
    {"name": "GPT 4.1", "deployment": "GPT 4.1"},
    {"name": "Gemini 2.5 Pro", "deployment": "Gemini-2.5-pro"},
    {"name": "Gemini 2.0 Flash", "deployment": "Gemini-2.0-flash"},
]

# ANSI color codes
class Colors:
    RESET = "\033[0m"
    RED = "\033[31m"
    GREEN = "\033[32m"
    YELLOW = "\033[33m"
    BLUE = "\033[34m"
    MAGENTA = "\033[35m"
    CYAN = "\033[36m"
    GRAY = "\033[90m"


def log(message: str, color: str = Colors.RESET):
    """Print colored message"""
    print(f"{color}{message}{Colors.RESET}")


def log_section(title: str):
    """Print section header"""
    print("\n" + "=" * 80)
    log(title, Colors.CYAN)
    print("=" * 80 + "\n")


def image_to_base64(image_path: str) -> str:
    """Convert image file to base64 string"""
    try:
        with open(image_path, "rb") as image_file:
            return base64.b64encode(image_file.read()).decode("utf-8")
    except Exception as e:
        raise Exception(f"Failed to read image file: {e}")


def download_image_to_base64(url: str, verify_ssl: bool = True) -> str:
    """Download image from URL and convert to base64"""
    try:
        response = requests.get(url, timeout=30, verify=verify_ssl)
        response.raise_for_status()
        return base64.b64encode(response.content).decode("utf-8")
    except Exception as e:
        raise Exception(f"Failed to download image from URL: {e}")


def generate_test_image() -> str:
    """Generate a simple test image (SVG) as base64"""
    svg = """<?xml version="1.0" encoding="UTF-8"?>
<svg width="400" height="300" xmlns="http://www.w3.org/2000/svg">
  <!-- Background -->
  <rect width="400" height="300" fill="#f0f0f0"/>

  <!-- Colored squares -->
  <rect x="50" y="50" width="80" height="80" fill="#FF6B6B" stroke="#000" stroke-width="2"/>
  <rect x="160" y="50" width="80" height="80" fill="#4ECDC4" stroke="#000" stroke-width="2"/>
  <rect x="270" y="50" width="80" height="80" fill="#45B7D1" stroke="#000" stroke-width="2"/>

  <rect x="105" y="160" width="80" height="80" fill="#FFA07A" stroke="#000" stroke-width="2"/>
  <rect x="215" y="160" width="80" height="80" fill="#98D8C8" stroke="#000" stroke-width="2"/>

  <!-- Text -->
  <text x="200" y="30" font-family="Arial" font-size="24" fill="#333" text-anchor="middle">
    Test Pattern
  </text>
</svg>"""
    return base64.b64encode(svg.encode("utf-8")).decode("utf-8")


def save_test_image(base64_data: str, output_path: str):
    """Save base64 image data to file"""
    with open(output_path, "wb") as f:
        f.write(base64.b64decode(base64_data))
    log(f"✅ Test image saved to: {output_path}", Colors.GREEN)


def test_vision_model(
    model: Dict[str, str],
    image_base64: str,
    image_type: str = "jpeg",
    api_key: str = "",
    verbose: bool = False,
    verify_ssl: bool = True,
) -> Dict[str, Any]:
    """Test vision analysis with a single model"""
    start_time = time.time()

    try:
        log(f"  Testing {model['name']}...", Colors.GRAY)

        # Prepare the vision request (OpenAI format)
        request = {
            "model": model["deployment"],
            "messages": [
                {
                    "role": "user",
                    "content": [
                        {
                            "type": "text",
                            "text": """Analyze this image in detail. Describe:
1. What you see in the image (objects, people, scene)
2. Colors and composition
3. Lighting conditions
4. Overall mood or atmosphere
5. Any text visible in the image

Be specific and detailed.""",
                        },
                        {
                            "type": "image_url",
                            "image_url": {
                                "url": f"data:image/{image_type};base64,{image_base64}",
                                "detail": "high",
                            },
                        },
                    ],
                }
            ],
            "max_tokens": 500,
            "temperature": 0.3,
        }

        headers = {"Content-Type": "application/json"}
        if api_key:
            headers["Authorization"] = f"Bearer {api_key}"

        response = requests.post(ENDPOINT, headers=headers, json=request, timeout=60, verify=verify_ssl)
        response_time = (time.time() - start_time) * 1000  # Convert to ms

        if not response.ok:
            error_data = response.json()
            return {
                "model": model["name"],
                "success": False,
                "response_time": response_time,
                "status_code": response.status_code,
                "error": error_data.get("error", {}).get("message")
                or error_data.get("message")
                or str(error_data),
            }

        data = response.json()

        # Handle both wrapped and unwrapped responses
        wrapped_data = data.get("data", data)
        content = wrapped_data.get("choices", [{}])[0].get("message", {}).get("content", "")
        tokens_used = wrapped_data.get("usage", {}).get("total_tokens")

        return {
            "model": model["name"],
            "success": True,
            "response_time": response_time,
            "status_code": response.status_code,
            "response": content,
            "tokens_used": tokens_used,
        }

    except Exception as e:
        response_time = (time.time() - start_time) * 1000
        return {
            "model": model["name"],
            "success": False,
            "response_time": response_time,
            "error": str(e),
        }


def main():
    """Main test function"""
    parser = argparse.ArgumentParser(description="Test Midas API vision capabilities")
    parser.add_argument("--image", help="Path to local image file")
    parser.add_argument("--url", help="URL to image file")
    parser.add_argument("--model", help="Test specific model only")
    parser.add_argument("--api-key", help="Midas API key (optional)")
    parser.add_argument("--verbose", "-v", action="store_true", help="Show full responses")
    parser.add_argument("--generate-test-image", action="store_true", help="Generate test pattern")
    parser.add_argument("--no-verify-ssl", action="store_true", help="Disable SSL verification (for corporate APIs)")

    args = parser.parse_args()

    log_section("🔍 Midas API Vision/Image Analysis Test")

    # Handle test image generation
    if args.generate_test_image:
        log_section("Generating Test Image")
        test_image_base64 = generate_test_image()
        output_path = Path(__file__).parent / "test-pattern.svg"
        save_test_image(test_image_base64, str(output_path))
        log("\nYou can now test with this image:", Colors.YELLOW)
        log(f"  python scripts/test-vision-analysis.py --image {output_path}", Colors.CYAN)
        return

    # Validate inputs
    if not args.image and not args.url:
        log("❌ Error: No image source provided!", Colors.RED)
        log("\nUsage:", Colors.YELLOW)
        log("  --image path/to/image.jpg    Load image from file", Colors.CYAN)
        log("  --url https://example.com/image.jpg    Load image from URL", Colors.CYAN)
        log("  --generate-test-image         Generate a test pattern SVG", Colors.CYAN)
        log('  --model "GPT 4o"             Test specific model only', Colors.CYAN)
        log("  --verbose                     Show full responses", Colors.CYAN)
        sys.exit(1)

    # Load image
    image_base64: str
    image_type = "jpeg"

    try:
        if args.image:
            log(f"Loading image from file: {args.image}", Colors.GRAY)
            image_base64 = image_to_base64(args.image)

            # Detect image type from extension
            ext = Path(args.image).suffix.lower()
            if ext == ".png":
                image_type = "png"
            elif ext == ".gif":
                image_type = "gif"
            elif ext == ".webp":
                image_type = "webp"
            elif ext == ".svg":
                image_type = "svg+xml"

            file_size_kb = len(base64.b64decode(image_base64)) // 1024
            log(f"✅ Image loaded: {file_size_kb} KB ({image_type})", Colors.GREEN)

        elif args.url:
            log(f"Downloading image from URL: {args.url}", Colors.GRAY)
            image_base64 = download_image_to_base64(args.url, verify_ssl=not args.no_verify_ssl)
            file_size_kb = len(base64.b64decode(image_base64)) // 1024
            log(f"✅ Image downloaded: {file_size_kb} KB", Colors.GREEN)

    except Exception as e:
        log(f"❌ Failed to load image: {e}", Colors.RED)
        sys.exit(1)

    # Filter models if specific model requested
    models_to_test = VISION_MODELS
    if args.model:
        models_to_test = [
            m for m in VISION_MODELS if m["name"] == args.model or m["deployment"] == args.model
        ]
        if not models_to_test:
            log(f"❌ Model not found: {args.model}", Colors.RED)
            log("Available models:", Colors.YELLOW)
            for m in VISION_MODELS:
                log(f"  - {m['name']}", Colors.CYAN)
            sys.exit(1)

    # Get API key
    api_key = args.api_key or API_KEY

    # Run tests
    log_section("Running Vision Analysis Tests")
    results: List[Dict[str, Any]] = []
    success_count = 0
    fail_count = 0

    # Disable SSL warnings if --no-verify-ssl is used
    if args.no_verify_ssl:
        import urllib3
        urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
        log("⚠️  SSL verification disabled", Colors.YELLOW)

    for model in models_to_test:
        result = test_vision_model(
            model, image_base64, image_type, api_key, args.verbose, verify_ssl=not args.no_verify_ssl
        )
        results.append(result)

        if result["success"]:
            success_count += 1
            log(f"  ✅ {result['model']} - {result['response_time']:.0f}ms", Colors.GREEN)
            if result.get("tokens_used"):
                log(f"     Tokens used: {result['tokens_used']}", Colors.GRAY)
            if args.verbose and result.get("response"):
                log(f"\n     Response:", Colors.CYAN)
                response_text = result["response"][:500]
                if len(result["response"]) > 500:
                    response_text += "..."
                log(f"     {response_text}", Colors.GRAY)
                log("", Colors.RESET)
        else:
            fail_count += 1
            log(f"  ❌ {result['model']} - {result['response_time']:.0f}ms", Colors.RED)
            if result.get("status_code"):
                log(f"     Status: {result['status_code']}", Colors.RED)
            if result.get("error"):
                log(f"     Error: {result['error'][:200]}", Colors.RED)

        # Small delay to avoid rate limiting
        time.sleep(1)

    # Summary
    log_section("Test Summary")
    log(f"Total tests: {len(results)}", Colors.CYAN)
    log(f"✅ Successful: {success_count}", Colors.GREEN)
    log(f"❌ Failed: {fail_count}", Colors.RED)
    if results:
        success_rate = (success_count / len(results)) * 100
        log(f"Success rate: {success_rate:.1f}%", Colors.CYAN)

    # Average response time for successful tests
    successful_results = [r for r in results if r["success"]]
    if successful_results:
        avg_time = sum(r["response_time"] for r in successful_results) / len(successful_results)
        log(f"Average response time: {avg_time:.0f}ms", Colors.CYAN)

    # Save results to file
    results_path = Path(__file__).parent / "vision-test-results.json"
    with open(results_path, "w") as f:
        json.dump(
            {
                "timestamp": datetime.now().isoformat(),
                "image_source": args.image or args.url,
                "image_type": image_type,
                "total_tests": len(results),
                "success_count": success_count,
                "fail_count": fail_count,
                "results": [
                    {
                        **r,
                        # Truncate response in saved file for readability
                        "response": r.get("response", "")[:500] if r.get("response") else None,
                    }
                    for r in results
                ],
            },
            f,
            indent=2,
        )
    log(f"\n📝 Results saved to: {results_path}", Colors.BLUE)

    # Show sample responses if not in verbose mode
    if not args.verbose and success_count > 0:
        log_section("Sample Responses (first 200 chars each)")
        for r in results:
            if r["success"] and r.get("response"):
                log(f"\n{r['model']}:", Colors.CYAN)
                log(r["response"][:200] + "...", Colors.GRAY)
        log("\n💡 Use --verbose flag to see full responses", Colors.YELLOW)

    print("\n")


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        log("Fatal error:", Colors.RED)
        print(str(e))
        sys.exit(1)
